pennylane/decomposition/decomposition_rule.py

```diff
@@ -0,0 +1,230 @@
+        def cnot_decomp(wires):
+            qml.H(wires=wires[1])
+            qml.CZ(wires=wires)
+            qml.H(wires=wires[1])
+
+        @cnot_decomp.resources
+        def _():
+            return {
+                qml.H: 2,
+                qml.CZ: 1,
+            }
+
+        qml.add_decomposition(qml.CNOT, cnot_decomp)
+
+    .. details::
+        :title: Usage Details
+
+        In many cases, the resource function of an operator's decomposition is not static. For
+        example, consider ``MultiRZ``, whose decomposition function looks like:
+
+        .. code-block:: python
+
+            import pennylane as qml
+
+            def _multi_rz_decomposition(theta, wires, **__):
+                for w0, w1 in zip(wires[-1:0:-1], wires[-2::-1]):
+                    qml.CNOT(wires=(w0, w1))
+                qml.RZ(theta, wires=wires[0])
+                for w0, w1 in zip(wires[1:], wires[:-1]):
+                    qml.CNOT(wires=(w0, w1))
+
+        We notice that the number of ``CNOT`` gates required to compose a ``MultiRZ`` depends on
+        the number of wires. For each operator class, we can find the set of parameters that
+        affect the gate count of its decompositions in the ``resource_param_keys`` attribute:
+
+        >>> qml.CNOT.resource_param_keys
+        {}
+        >>> qml.MultiRZ.resource_param_keys
+        {'num_wires'}
+
+        The resource function for any decomposition of ``MultiRZ`` should take ``num_wires`` as
+        an argument:
+
+        .. code-block:: python
+
+            def _multi_rz_decomposition_resources(num_wires):
+                return {
+                    qml.CNOT: 2 * (num_wires - 1),
+                    qml.RZ: 1
+                }
+
+        Consequentially, two ``MultiRZ`` gates acting on different numbers of wires will have
+        different decompositions. As a result, when defining a decomposition rule that contains
+        ``MultiRZ`` gates, we require that more information is provided.
+
+        Consider a fictitious operator with the following decomposition:
+
+        .. code-block:: python
+
+            def _my_decomposition(thata, wires):
+                qml.MultiRZ(theta, wires=wires[:-1])
+                qml.MultiRZ(theta, wires=wires)
+                qml.MultiRZ(theta, wires=wires[1:])
+
+        It contains two ``MultiRZ`` gates on ``num_wires - 1`` wires and one ``MultiRZ`` gate on
+        ``num_wires`` wires, which is reflected in its corresponding resource function:
+
+        .. code-block:: python
+
+            def _my_decomposition_resources(num_wires):
+                return {
+                    qml.make_resource_rep(qml.MultiRZ, num_wires=num_wires - 1): 2,
+                    qml.make_resource_rep(qml.MultiRZ, num_wires=num_wires): 1
+                }
+
+        where ``qml.make_resource_rep`` is a utility function that wraps an operator type and
+        any additional information relavent to its resource estimate into a data structure.
```

Comment:
I find this narrative easier to understand. Following this narrative, maybe we can give this function a different name, completely hiding away the concept of a resource rep or `CompressedResourceOp` from the user?